### DISCUSSION TITLE
Allow changing the Task comments via a task

### DISCUSSION THREAD

**Discussion by xeor** [apache/airflow#28448] [General]
### Description

A way to set the new task-comment that arrived in 2.5 using a task.

### Use case/motivation

I want to summarize a dag-run with some additional information from the context available inside the dag-run. Example with links and data that should be easy to look up.

This will work great with https://github.com/apache/airflow/issues/28167 to add markdown support for adding links from inside a task.

Something like

```py
    @task(task_id="summarizer")
    def preflight(ti=None):
        ti.notes = 'some notes with context'
```

### Related issues

https://github.com/apache/airflow/discussions/28035

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
IMHO - This should be done via API not via decorator/task. There is no "internal" mechanism to change some tasks from other tasks and stable REST API is the best way of doing it. Let me convert it to a discussion - maybe others will see it differently but I have a feeling that the way it is specified (what should be decorateD? when? when it should be processed? etc. is far from "defined". This is more a wish than defined feature, and discussion is more appropriate place for it.

**Reply by xeor** [reply]
The usecase for me is to populate some links and a basic report inside the dag notes. I didnt know that there are both dag-notes and task-notes, but my usecase is for dag-notes.

So an updated example of what we need would be

```py
def summarizer(context):
    context.notes = '1234 files scanned. You can find the logs at ..../?dag={context["run_id"]}'

with DAG(
    on_success_callback=summarizer
    ....
) as dag:
```

I'm not sure if there is a way for summarizer to have access to change the dag info tho. Or if this is what you ment by using the api?

**Reply by potiuk** [reply]
This is the API https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/set_mapped_task_instance_note

Initially task instances were foreseen for Manual entry by users (see the comment in the API). This is something that users can do via UI, that's why we have also API for that. And you can use it even now - you just need to use the API, authorize as dedicated users and you can manage it even today. This is why I mentioned it - because it solves your problem immediately. The feature is there - ready to be used also for your tasks.

But If you think it would make sense to implement some internal ways of creating such notes and basically modifying notes without going through public APIk this is a bit different case - you need to define some kind of internal API (for example Configurations and Variables are this kind of internal API) and make it "stable" - i.e. not only oimplement it but also make sure it can be future-proof (our current stable REST API serves that purpose - that's why it's good to use it for any kind of updates like that - even if it is more of a hassle - it is future-proof). 

However, any proposal of introducing a new "kind" of API that can be used by DAG authors should be something that we take lightly - this should be formally proposed at the devlist so that the community members agree on adding such a "future maintained" API for DAG authors.

Currently you **could** if you would really like to do it modify directly the DB of Airflow and create such note. But this is not something you can rely on in the future versions of Airflow - that's why the REST API makes much more sense.

**Reply by xeor** [reply]
Got the base working :) Thanks for the feedback! Will polish before use, but the gist is (for others trying to do the same

```py
def success_callback(context):
    import json
    import base64
    import requests

    username = 'airflow'
    password = 'airflow'
    auth_bytes = f'{username}:{password}'.encode()
    base64_bytes = base64.b64encode(auth_bytes)
    base64_string = base64_bytes.decode()

    data = {'note': 'test'}

    url = f'http://airflow-webserver:8080/api/v1/dags/{context["dag"].dag_id}/dagRuns/{context["dag_run"].run_id}/setNote'

    api_request = requests.patch(
        url,
        data=json.dumps(data),
        headers={
            "Authorization": f"Basic {base64_string}",
            "Content-Type": "application/json",
        },
    )

with models.DAG(
    dag_id="callbacktesting",
    schedule=None,
    start_date=pendulum.datetime(2022, 1, 1, tz="UTC"),
    on_success_callback=success_callback,
) as dag:

    @task(task_id="test1")
    def preflight():
        print("hello")
    preflight_task = preflight()
```
